velox/exec/tests/HashJoinTest.cpp

```diff
@@ -711,483 +819,1202 @@ TEST_F(HashJoinTest, arrayBasedLookup) {
+        .joinOutputLayout({"c1"})
+        .referenceQuery(
+            "SELECT t.c1 FROM t WHERE t.c0 NOT IN (SELECT c0 FROM u WHERE c0 IS NOT NULL)")
+        .run();
+  }
 
-  auto planNodeIdGenerator = std::make_shared<PlanNodeIdGenerator>();
-  auto op = PlanBuilder(planNodeIdGenerator)
-                .values({leftVectors})
-                .hashJoin(
-                    {"t0"},
-                    {"u0"},
-                    PlanBuilder(planNodeIdGenerator)
-                        .values({rightVectors})
-                        .planNode(),
-                    "",
-                    {"t0", "t1"},
-                    core::JoinType::kLeftSemi)
-                .planNode();
-
-  assertQuery(
-      op, "SELECT t.* FROM t WHERE EXISTS (SELECT u0 FROM u WHERE t0 = u0)");
-
-  op = PlanBuilder(planNodeIdGenerator)
-           .values({leftVectors})
-           .hashJoin(
-               {"t0"},
-               {"u0"},
-               PlanBuilder(planNodeIdGenerator)
-                   .values({rightVectors})
-                   .planNode(),
-               "t1 != u1",
-               {"t0", "t1"},
-               core::JoinType::kLeftSemi)
-           .planNode();
-
-  assertQuery(
-      op,
-      "SELECT t.* FROM t WHERE EXISTS (SELECT u0, u1 FROM u WHERE t0 = u0 AND t1 <> u1)");
+  // Empty build side.
+  {
+    auto testProbeVectors = probeVectors;
+    auto testBuildVectors = buildVectors;
+    HashJoinBuilder joinBuilder(*pool_, duckDbQueryRunner_);
+    joinBuilder.numDrivers(numDrivers_)
+        .probeKeys({"c0"})
+        .probeVectors(std::move(testProbeVectors))
+        .buildKeys({"c0"})
+        .buildVectors(std::move(testBuildVectors))
+        .buildFilter("c0 < 0")
+        .joinType(core::JoinType::kNullAwareAnti)
+        .joinOutputLayout({"c1"})
+        .referenceQuery(
+            "SELECT t.c1 FROM t WHERE t.c0 NOT IN (SELECT c0 FROM u WHERE c0 < 0)")
+        .run();
+  }
+
+  // Build side with nulls. Null-aware Anti join always returns nothing.
+  {
+    auto testProbeVectors = probeVectors;
+    auto testBuildVectors = buildVectors;
+    HashJoinBuilder joinBuilder(*pool_, duckDbQueryRunner_);
+    joinBuilder.numDrivers(numDrivers_)
+        .probeKeys({"c0"})
+        .probeVectors(std::move(testProbeVectors))
+        .buildKeys({"c0"})
+        .buildVectors(std::move(testBuildVectors))
+        .joinType(core::JoinType::kNullAwareAnti)
+        .joinOutputLayout({"c1"})
+        .referenceQuery(
+            "SELECT t.c1 FROM t WHERE t.c0 NOT IN (SELECT c0 FROM u)")
+        .run();
+  }
 }
 
-TEST_F(HashJoinTest, rightSemiJoin) {
-  // leftVector size is greater than rightVector size.
-  auto leftVectors = makeRowVector(
-      {"u0", "u1"},
-      {
-          makeFlatVector<int32_t>(
-              1'234, [](auto row) { return row % 11; }, nullEvery(13)),
-          makeFlatVector<int32_t>(1'234, [](auto row) { return row; }),
+TEST_P(MultiThreadedHashJoinTest, nullAwareAntiJoinWithFilter) {
```

Comment:
There are 2 tests `nullAwareAntiJoinWithFilterAndNullKey` and `nullAwareAntiJoinWithFilterOnNullableColumn` I added in #2583, can we add them here as well?

Reply:
Thanks and included.